zkevm-circuits/src/evm_circuit/util/math_gadget/addwords.rs

```diff
@@ -0,0 +1,242 @@
+        &self.addends
+    }
+
+    pub(crate) fn sum(&self) -> &util::Word<F> {
+        &self.sum
+    }
+
+    pub(crate) fn carry(&self) -> &Option<util::Cell<F>> {
+        &self.carry_hi
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::super::test_util::*;
+    use super::*;
+    use eth_types::Word;
+    use halo2_proofs::halo2curves::bn256::Fr;
+    use halo2_proofs::plonk::Error;
+
+    #[test]
+    fn test_addwords() {
+        #[derive(Clone)]
+        /// sum = a + b
+        struct AddWordsTestContainer<F> {
+            addwords_gadget: AddWordsGadget<F, 2, true>,
+            a: util::Word<F>,
+            b: util::Word<F>,
+            sum: util::Word<F>,
+        }
+
+        impl<F: Field> MathGadgetContainer<F> for AddWordsTestContainer<F> {
+            const NAME: &'static str = "AddWordsGadget";
+
+            fn configure_gadget_container(cb: &mut ConstraintBuilder<F>) -> Self {
+                let a = cb.query_word();
+                let b = cb.query_word();
+                let sum = cb.query_word();
+                let addwords_gadget = AddWordsGadget::<F, 2, true>::construct(
+                    cb,
+                    [a.clone(), b.clone()],
+                    sum.clone(),
+                );
+                AddWordsTestContainer {
+                    addwords_gadget,
+                    a,
+                    b,
+                    sum,
+                }
+            }
+
+            fn assign_gadget_container(
+                &self,
+                input_words: &[Word],
+                region: &mut CachedRegion<'_, '_, F>,
+            ) -> Result<(), Error> {
+                let a = input_words[0];
+                let b = input_words[1];
+                let sum = input_words[2];
+                let offset = 0;
+
+                self.a.assign(region, offset, Some(a.to_le_bytes()))?;
+                self.b.assign(region, offset, Some(b.to_le_bytes()))?;
+                self.sum.assign(region, offset, Some(sum.to_le_bytes()))?;
+                self.addwords_gadget.assign(region, 0, [a, b], sum)?;
+
+                Ok(())
+            }
+        }
+
+        test_math_gadget_container::<Fr, AddWordsTestContainer<Fr>>(
+            vec![Word::from(0), Word::from(0), Word::from(0)],
+            true,
+        );
+
+        test_math_gadget_container::<Fr, AddWordsTestContainer<Fr>>(
+            vec![Word::from(1), Word::from(1), Word::from(2)],
+            true,
+        );
+
+        test_math_gadget_container::<Fr, AddWordsTestContainer<Fr>>(
+            vec![Word::from(1000), Word::from(1000), Word::from(2000)],
+            true,
+        );
+
+        test_math_gadget_container::<Fr, AddWordsTestContainer<Fr>>(
+            vec![Word::MAX - 1, Word::from(1), Word::MAX],
+            true,
+        );
+
+        test_math_gadget_container::<Fr, AddWordsTestContainer<Fr>>(
+            vec![Word::MAX, Word::from(1), Word::from(0)],
+            false,
+        );
```

Comment:
A test like this without `CHECK_OVREFLOW` enabled would interesting as well.

Also some tests with `N_ADDENDS` set to something `> 2` would be good.

Reply:
Sure, will add more